File: src/test/scala/com/exasol/cloudetl/kafka/JsonArgumentMatcher.scala
package com.exasol.cloudetl.kafka

import com.exasol.common.json.JsonMapper

import com.fasterxml.jackson.databind.JsonNode
import org.mockito.ArgumentMatcher
import org.scalactic.TypeCheckedTripleEquals.convertToCheckingEqualizer
import org.scalatest.Assertions.{===, unconstrainedEquality}

/**
 * The order of JSON fields is not deterministic, therefore, we need a custom
 * matchers that compares the JSON object field by field.
 */
class JsonArgumentMatcher(expectedJson: String) extends ArgumentMatcher[String] {

  private[this] val expectedJsonNode = JsonMapper.fromJson[JsonNode](expectedJson)

  override final def matches(argument: String): Boolean =
    JsonMapper.fromJson[JsonNode](argument) === expectedJsonNode
}

File: src/test/java/com/exasol/cloudetl/kafka/JsonArgumentMatcher.java
package com.exasol.cloudetl.kafka;

import org.mockito.ArgumentMatcher;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * The order of JSON fields is not deterministic, therefore this matcher compares parsed JSON trees.
 */
public final class JsonArgumentMatcher implements ArgumentMatcher<String> {
    private static final ObjectMapper JSON = new ObjectMapper();
    private final JsonNode expectedJsonNode;

    public JsonArgumentMatcher(final String expectedJson) {
        this.expectedJsonNode = readJson(expectedJson);
    }

    @Override
    public boolean matches(final String argument) {
        return readJson(argument).equals(this.expectedJsonNode);
    }

    public static JsonNode readJson(final String json) {
        try {
            return JSON.readTree(json);
        } catch (final JsonProcessingException exception) {
            throw new IllegalArgumentException("Invalid JSON: " + json, exception);
        }
    }
}
